framework/audio/engine/platform/lin/linuxaudiodriver.cpp

```diff
@@ -0,0 +1,162 @@
+#include "linuxaudiodriver.h"
+
+#define ALSA_PCM_NEW_HW_PARAMS_API
+#include <alsa/asoundlib.h>
+
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <math.h>
+#include <pthread.h>
+
+#include "log.h"
+
+using namespace mu::audio::engine;
+
+namespace  {
+struct ALSAData
+{
+    float* buffer;
+    snd_pcm_t* alsaDeviceHandle;
+    int samples;
+    int channels;
+    bool audioProcessingDone;
+    pthread_t threadHandle;
+    IAudioDriver::Callback callback;
+    void* userdata;
+};
+
+static ALSAData* _alsaData{ nullptr };
+
+static void* alsaThread(void* aParam)
+{
+    ALSAData* data = static_cast<ALSAData*>(aParam);
+
+    int ret = snd_pcm_wait(data->alsaDeviceHandle, 1000);
+    IF_ASSERT_FAILED(ret > 0) {
+        return nullptr;
+    }
+
+    while (!data->audioProcessingDone)
+    {
+        uint8_t* stream = (uint8_t*)data->buffer;
+        int len = data->samples * data->channels * sizeof(float);
+
+        data->callback(data->userdata, stream, len);
+
+        snd_pcm_sframes_t pcm = snd_pcm_writei(data->alsaDeviceHandle, data->buffer, data->samples);
+        if (pcm != -EPIPE) {
+        } else {
+            snd_pcm_prepare(data->alsaDeviceHandle);
+        }
+    }
+
+    return nullptr;
+}
+
+static void alsaCleanup()
+{
+    IF_ASSERT_FAILED(_alsaData) {
+        return;
+    }
+
+    _alsaData->audioProcessingDone = true;
```

Comment:
Are you going to alter it according to our code-style in the future?

Reply:
By the way, how to designate a global variable (within the file), i.e. this is not a member of the class.